File: Back-End/backend/src/global/rateHelper.ts
import { MarketIndex } from '../models/MarketIndex';
import { getOrSet } from './cache';

/**
 * 取得即時匯率 Map（code → 台幣匯率），TWD 固定為 1
 * 使用 market:forex-rates 快取（TTL 300s）
 */
export async function getLiveRateMap(): Promise<Record<string, number | null>> {
  const rates = await getOrSet(
    'market:forex-rates',
    () => MarketIndex.fetchForexRates(),
    300
  );
  const map: Record<string, number | null> = { TWD: 1 };
  for (const r of rates) map[r.code] = r.rate;
  return map;
}

File: Back-End/python-backend/services/rate_helper.py
from services.yahoo_finance import get_forex_rates


def get_live_rate_map() -> dict[str, float | None]:
    """回傳 {幣別代碼: 台幣匯率} dict，TWD 固定為 1"""
    rates = get_forex_rates()
    rate_map: dict[str, float | None] = {"TWD": 1.0}
    for r in rates:
        rate_map[r["code"]] = r["rate"]
    return rate_map
